File: app/src/main/java/com/lumiyaviewer/lumiya/slproto/messages/FetchInventoryDescendents.java
package com.lumiyaviewer.lumiya.slproto.messages;

import com.lumiyaviewer.lumiya.slproto.SLMessage;
import java.nio.ByteBuffer;
import java.util.UUID;

/**
 * Get inventory segment.
 *
 * <p>Template: {@code FetchInventoryDescendents Low 277 NotTrusted Zerocoded}
 * (recovered/reference/message_template.msg).
 */
public class FetchInventoryDescendents extends SLMessage {
    public AgentData AgentData_Field;
    public InventoryData InventoryData_Field;

    /** Block AgentData, Single. */
    public static class AgentData {
        public UUID AgentID; // LLUUID
        public UUID SessionID; // LLUUID
    }

    /** Block InventoryData, Single. */
    public static class InventoryData {
        public boolean FetchFolders; // BOOL - false will omit folders in query
        public boolean FetchItems; // BOOL - false will omit items in query
        public UUID FolderID; // LLUUID
        public UUID OwnerID; // LLUUID
        public int SortOrder; // S32 - 0 = name, 1 = time
    }

    public FetchInventoryDescendents() {
        this.zeroCoded = true;
        this.AgentData_Field = new AgentData();
        this.InventoryData_Field = new InventoryData();
    }

    @Override
    public int CalcPayloadSize() {
        return 74;
    }

    @Override
    public void Handle(SLMessageHandler messageHandler) {
        messageHandler.HandleFetchInventoryDescendents(this);
    }

    @Override
    public void PackPayload(ByteBuffer byteBuffer) {
        // Message number: Low 277 (FetchInventoryDescendents).
        byteBuffer.putShort((short) 0xFFFF);
        byteBuffer.put((byte) 0x01);
        byteBuffer.put((byte) 0x15);
        packUUID(byteBuffer, this.AgentData_Field.AgentID);
        packUUID(byteBuffer, this.AgentData_Field.SessionID);
        packUUID(byteBuffer, this.InventoryData_Field.FolderID);
        packUUID(byteBuffer, this.InventoryData_Field.OwnerID);
        packInt(byteBuffer, this.InventoryData_Field.SortOrder);
        packBoolean(byteBuffer, this.InventoryData_Field.FetchFolders);
        packBoolean(byteBuffer, this.InventoryData_Field.FetchItems);
    }

    @Override
    public void UnpackPayload(ByteBuffer byteBuffer) {
        this.AgentData_Field.AgentID = unpackUUID(byteBuffer);
        this.AgentData_Field.SessionID = unpackUUID(byteBuffer);
        this.InventoryData_Field.FolderID = unpackUUID(byteBuffer);
        this.InventoryData_Field.OwnerID = unpackUUID(byteBuffer);
        this.InventoryData_Field.SortOrder = unpackInt(byteBuffer);
        this.InventoryData_Field.FetchFolders = unpackBoolean(byteBuffer);
        this.InventoryData_Field.FetchItems = unpackBoolean(byteBuffer);
    }
}

File: app/src/main/java/com/lumiyaviewer/lumiya/slproto/messages/FetchInventoryDescendents.kt
package com.lumiyaviewer.lumiya.slproto.messages

import com.lumiyaviewer.lumiya.slproto.SLMessage
import java.nio.ByteBuffer
import java.util.UUID

/**
 * Get inventory segment.
 *
 * <p>Template: {@code FetchInventoryDescendents Low 277 NotTrusted Zerocoded}
 * (recovered/reference/message_template.msg).
 */
open class FetchInventoryDescendents : SLMessage() {
    @JvmField var AgentData_Field: AgentData = AgentData()
    @JvmField var InventoryData_Field: InventoryData = InventoryData()

    /** Block AgentData, Single. */
    open class AgentData {
        @JvmField var AgentID: UUID? = null
        @JvmField var SessionID: UUID? = null
    }

    /** Block InventoryData, Single. */
    open class InventoryData {
        @JvmField var FetchFolders: Boolean = false
        @JvmField var FetchItems: Boolean = false
        @JvmField var FolderID: UUID? = null
        @JvmField var OwnerID: UUID? = null
        @JvmField var SortOrder: Int = 0
    }

    init {
        zeroCoded = true
    }

    override fun CalcPayloadSize(): Int {
        return 74
    }

    override fun Handle(messageHandler: SLMessageHandler) {
        messageHandler.HandleFetchInventoryDescendents(this)
    }

    override fun PackPayload(byteBuffer: ByteBuffer) {
        // Message number: Low 277 (FetchInventoryDescendents).
        byteBuffer.putShort(0xFFFF.toShort())
        byteBuffer.put((0x01).toByte())
        byteBuffer.put((0x15).toByte())
        packUUID(byteBuffer, AgentData_Field.AgentID)
        packUUID(byteBuffer, AgentData_Field.SessionID)
        packUUID(byteBuffer, InventoryData_Field.FolderID)
        packUUID(byteBuffer, InventoryData_Field.OwnerID)
        packInt(byteBuffer, InventoryData_Field.SortOrder)
        packBoolean(byteBuffer, InventoryData_Field.FetchFolders)
        packBoolean(byteBuffer, InventoryData_Field.FetchItems)
    }

    override fun UnpackPayload(byteBuffer: ByteBuffer) {
        AgentData_Field.AgentID = unpackUUID(byteBuffer)
        AgentData_Field.SessionID = unpackUUID(byteBuffer)
        InventoryData_Field.FolderID = unpackUUID(byteBuffer)
        InventoryData_Field.OwnerID = unpackUUID(byteBuffer)
        InventoryData_Field.SortOrder = unpackInt(byteBuffer)
        InventoryData_Field.FetchFolders = unpackBoolean(byteBuffer)
        InventoryData_Field.FetchItems = unpackBoolean(byteBuffer)
    }
}
